### PR TITLE
[nn] module: full_backward_pre_hook

### LINKED ISSUE #42824: [feature request] Pre-backward hooks for runtime profiling use case
Would it be possible to get a pre-backward hook, in addition to pre-forward, forward and pre-backward?

**My use case and current solution**
I want to profile the per-layer runtime of my model to identify the slowest layers. How I currently do this:

1. Assign a unique name to every leaf layer.
2. Register forward and backward hooks on every leaf layer of the model.
3. Torch.cuda.synchronize() and log the timestamp at which the hook for each layer is called.
4. Take the difference between subsequent timestamps in the log.

This gives me accurate per-layer computation time. However, the torch.cuda.synchronize() after each layer does not reflect the actual time the forward and backward prop takes.

**What I'd like to do instead**
I'd like to use torch.cuda.Events instead. This is easy to do fr the forward prop:
1. Have a start event in the pre-forward hook for each layer.
2. Have an end event in the forward hook for each layer.
3. Measure the elapsed time between the 2 events.

However, I can't do this for backprop because there isn't a pre-backward hook. Having access to pre-backward hooks will enable accurate runtime profiling for backprop.

Also, please let me know if there are alternatives to get accurate layer-wise runtime profiling using any other methods that don't have a reliance on a pre-backward hook.

cc @ezyang @gchanan @zou3519 @albanD @mruberry @jbschlosser @walterddr @kshitij12345 @saketh-are @robieta @chaekit @aaronenyeshi @ngimel @nbcsm @guotuofeng @guyang3532 @gaoteng-git @tiffzhaofb @ilia-cher @gdankel @bitfort @SsnL @gqchen

### PR DESCRIPTION
**Pull request by kshitij12345**
Fixes https://github.com/pytorch/pytorch/issues/42824

* [x] Test
* [x] Doc

### CODE REVIEW

File: torch/nn/modules/module.py [hunk 1]
```diff
@@ -155,7 +157,7 @@ def register_module_forward_hook(hook: Callable[..., None]) -> RemovableHandle:
 
 
 def register_module_backward_hook(
-    hook: Callable[['Module', _grad_t, _grad_t], Union[None, Tensor]]
+    hook: Callable[['Module', _grad_t, _grad_t], Union[None, _grad_t]]
```
Comment (kshitij12345) [inline torch/nn/modules/module.py:160]: I think the signature here was wrong.

Note: This change is not related to `backward_pre_hook`

File: torch/nn/modules/module.py [hunk 2]
```diff
@@ -181,8 +183,45 @@ def register_module_backward_hook(
...
+
+    """
+    handle = hooks.RemovableHandle(_global_backward_pre_hooks)
+    _global_backward_pre_hooks[handle.id] = hook
+    return handle
+
+
 def register_module_full_backward_hook(
-    hook: Callable[['Module', _grad_t, _grad_t], Union[None, Tensor]]
+    hook: Callable[['Module', _grad_t, _grad_t], Union[None, _grad_t]]
```
Comment (kshitij12345) [inline torch/nn/modules/module.py:224]: I think the signature here was wrong.

Note: This change is not related to `backward_pre_hook`

File: docs/source/notes/modules.rst [hunk 3]
```diff
@@ -603,10 +603,12 @@ PyTorch provides two types of hooks for modules:
   :func:`~torch.nn.modules.module.register_module_forward_pre_hook` and
   :func:`~torch.nn.modules.module.register_module_forward_hook` functions.
 * **Backward hooks** are called during the backward pass. They can be installed with
-  :func:`~torch.nn.Module.register_full_backward_hook`. These hooks will be called when the backward for this
-  Module has been computed and will allow the user to access the gradients for both the inputs and outputs.
+  :func:`~torch.nn.Module.register_backward_pre_hook` and :func:`~torch.nn.Module.register_full_backward_hook`.
```
Comment (albanD) [inline docs/source/notes/modules.rst:606]: can we keep the "full" in the name.
It is a bit weird but that improves consistency.

**Review by albanD — COMMENTED** [review]
@soulitzer can you take a look at this?

File: torch/nn/modules/module.py [hunk 4]
```diff
@@ -181,8 +183,45 @@ def register_module_backward_hook(
...
+    hook: Callable[['Module', _grad_t], Union[None, _grad_t]]
+) -> RemovableHandle:
+    r"""Registers a backward pre-hook common to all the modules.
+
+    .. warning ::
+        This adds global state to the `nn.module` module
+        and it is only intended for debugging/profiling purposes.
+
+    The hook will be called every time before the gradients with respect
+    to module inputs are computed. The hook should have the following signature::
```
Comment (soulitzer) [inline torch/nn/modules/module.py:196]: Could you reword as: "The hook will be called every time the gradients for the module are computed."

File: torch/nn/modules/module.py [hunk 5]
```diff
@@ -986,8 +1027,42 @@ def convert(t):
 
         return self._apply(convert)
 
+    def register_backward_pre_hook(
+        self, hook: Callable[['Module', _grad_t], Union[None, _grad_t]]
+    ) -> RemovableHandle:
+        r"""Registers a backward pre hook on the module.
+
+        The hook will be called every time before the gradients with respect
```
Comment (soulitzer) [inline torch/nn/modules/module.py:1035]: Same as above

File: torch/nn/modules/module.py [hunk 6]
```diff
@@ -986,8 +1027,42 @@ def convert(t):
...
+        to module inputs are computed. The hook should have the following signature::
+
+            hook(module, grad_output) -> Tensor or None
+
+        The :attr:`grad_output` is a tuple. The hook should
+        not modify its arguments, but it can optionally return a new gradient with
+        respect to the output that will be used in place of :attr:`grad_output` in
+        subsequent computations. Entry in :attr:`grad_output` will be ``None`` for
```
Comment (soulitzer) [inline torch/nn/modules/module.py:1043]: Same as above

File: torch/nn/modules/module.py [hunk 7]
```diff
@@ -986,8 +1027,42 @@ def convert(t):
 
         return self._apply(convert)
 
+    def register_backward_pre_hook(
+        self, hook: Callable[['Module', _grad_t], Union[None, _grad_t]]
+    ) -> RemovableHandle:
+        r"""Registers a backward pre hook on the module.
```
Comment (soulitzer) [inline torch/nn/modules/module.py:1033]: pre hook -> pre-hook

File: torch/nn/modules/module.py [hunk 8]
```diff
@@ -1248,6 +1335,8 @@ def __setstate__(self, state):
             self._non_persistent_buffers_set = set()
         if '_is_full_backward_hook' not in self.__dict__:
             self._is_full_backward_hook = None
+        if '_backward_pre_hooks' not in self.__dict__:
+            self._backward_pre_hooks = OrderedDict()
```
Comment (soulitzer) [inline torch/nn/modules/module.py:1339]: Hmm why don't I see _backward_hooks here?

File: docs/source/notes/modules.rst (hunk 3 above)
Comment (soulitzer) [reply docs/source/notes/modules.rst:606]: hmm I'm curious what the "full" means in the name

File: docs/source/notes/modules.rst [hunk 9]
```diff
@@ -603,10 +603,12 @@ PyTorch provides two types of hooks for modules:
...
+  These hooks will be called when the backward for this Module has been computed and will allow the user to
+  access the gradients for both the inputs and outputs.
```
Comment (soulitzer) [inline docs/source/notes/modules.rst:608]: we should update to say that you'll only see gradient outputs for register_backward_pre_hook

**Review by soulitzer — COMMENTED** [review]
Thanks, look pretty good. Just have some small comments on the docs.

File: torch/nn/modules/module.py (hunk 8 above)
Comment (kshitij12345) [reply torch/nn/modules/module.py:1339]: I think `_backward_hooks` has existed long before (atleast since 1.0.1 based on test_conv_backcompat in test_nn) and hence old checkpoints will probably have that field.

Without the update, these tests fail from test_nn: test_module_backcompat, test_conv_backcompat

File: test/test_nn.py [hunk 10]
```diff
@@ -326,6 +336,31 @@ def hook(module, grad_inputs, grad_outputs):
         output = bn(torch.randn(5, 5, requires_grad=True))
         output.sum().backward()
 
+    @skipIfTorchDynamo("TorchDynamo does not work well with hooks")
```
Comment (kshitij12345) [inline test/test_nn.py:339]: Fails on Dynamo. Have added skip similar to other hook related tests.

**Comment by kshitij12345** [comment]
@soulitzer have addressed all the points. PTAL Thanks :)

File: torch/nn/modules/module.py (hunk 8 above)
Comment (soulitzer) [reply torch/nn/modules/module.py:1423]: Gotcha, thanks

File: docs/source/notes/modules.rst (hunk 3 above)
Comment (albanD) [reply docs/source/notes/modules.rst]: Nothing. We just needed a different name than the current `register_backward_hook()` when the new version was added.
One day maybe we'll deprecate the old ones that are completely broken and use the full ones by default :dream:

**Review by soulitzer — APPROVED** [review]
LGTM

File: test/test_nn.py [hunk 11]
```diff
@@ -326,6 +336,31 @@ def hook(module, grad_inputs, grad_outputs):
...
+    def test_hook_backward_pre_and_full(self):
```
[the remark this answers is not in the document]
Comment (kshitij12345) [reply test/test_nn.py:340]: Makes sense. Renaming it to `test_backward_hook_interaction`
